### ISSUE TITLE
Resource server requests result in 403 (insufficient_access)

### ISSUE THREAD

**Issue by Thodor12** [openiddict/openiddict-core#1497]
### Confirm you've already contributed to this project or that you sponsor it

- [X] I confirm I'm a sponsor or a contributor

### Version

3.x

### Question

Hello, first off thanks for making this amazing library. I'm running into an issue where my resource server is constantly giving back errors saying insufficient_access.

I get the following log (enabled trace logging):
```
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (2ms) [Parameters=[@__identifier_0='?' (Size = 4000)], CommandType='Text', CommandTimeout='30']
      SELECT TOP(1) [c].[Id], [c].[Name], [c].[TenantIdentifier]
      FROM [CRM_Company] AS [c]
      WHERE [c].[TenantIdentifier] = @__identifier_0
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessRequestContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+InferIssuerFromHost.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+ExtractAccessTokenFromAuthorizationHeader.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+ExtractAccessTokenFromBodyForm.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+ExtractAccessTokenFromQueryString.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+ValidateToken.
trce: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The ASP.NET Core Data Protection token 'CfDJ8OpP9CaZdrhJkH68OTXlDjysdPNz2QHn-sGvfq4CWi8XFygCCaOPUO3an6nR0e5hRgQNglVRf1XDM8WenPn3PlSGu97v3ecOjBXT3O8nU7dRn9WWu18yshJEuflKgGwM9zNpyy7KETAxGSGmhWQl7N103HV47KulgAZavtNvS12m6WfLzCJrfe3N0cuWamr6nKKR6MNSv3g1da4Kxo53EtT2UeKUaCl_I21K-KsOnFwubWpfKzuCELkF4LaM_1qDFObvKKRb3Itl2o8FlDeTJmlEPoBA0G3chi_XuC8L2HjahqkY_mcLHGY15OWfeYuQdDbV9DnpCKscg8Bufj56_lszn2KNHzabkldZIE5lVMC318Vx7-_S5L_nLwv0J69mVkQzmvjX9Sr8oT8YrFNpu6b7IZV9Jd5AeyI6CcyXniGovv14nLA2u6-qYqtsrZCkgx-MJ1hgjrg83ykDTknaxg09bpcZUOBTQusW9JLhcEd0cyyGgykMsSFE271sYoHS1FkPFy8XL_NuFzqMwwwadePE14Fu6WUpUEr_RltsZB-2BSXUohq4fVF2KF8DMgbMSjslBoAQbf8bUvYK-eFKhTaObhATNh7sGf9N6JklwcWaEAq8uKSlVbn25iPRfkrieEU6kecwbMT5uwt94njWYS04-LNzU4WJ9LT9ZfeMWzt9D1Qs5b9p95oY5mBdcobn1LPu_qAjOng3qmwfjJNBpqE' was successfully validated and the following claims could be extracted: sub: 5, email: thom.vandenakker@outlook.com, name: Admin, role: Admin, client_id: mobile-app, oi_prst: mobile-app, oi_scp: openid, oi_scp: profile, oi_scp: email, oi_scp: roles, oi_scp: offline_access, oi_scp: mobile.access, oi_au_id: 874, oi_crt_dt: Thu, 04 Aug 2022 18:15:34 GMT, oi_exp_dt: Thu, 04 Aug 2022 18:18:34 GMT, oi_tkn_id: 857, oi_tkn_typ: access_token.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.DataProtection.OpenIddictValidationDataProtectionHandlers+ValidateDataProtectionToken.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+IntrospectToken.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+NormalizeScopeClaims.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+MapInternalClaims.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+ValidatePrincipal.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+ValidateExpirationDate.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+ValidateAudience.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+AttachHostChallengeError.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+AttachDefaultChallengeError.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+AttachHttpResponseCode`1[[OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext, OpenIddict.Validation, Version=3.1.1.0, Culture=neutral, PublicKeyToken=35a561290d20de2f]].
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+AttachCacheControlHeader`1[[OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext, OpenIddict.Validation, Version=3.1.1.0, Culture=neutral, PublicKeyToken=35a561290d20de2f]].
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+AttachWwwAuthenticateHeader`1[[OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext, OpenIddict.Validation, Version=3.1.1.0, Culture=neutral, PublicKeyToken=35a561290d20de2f]].
info: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The response was successfully returned as a challenge response: {
        "error": "insufficient_access",
        "error_description": "The user represented by the token is not allowed to perform the requested action.",
        "error_uri": "https://documentation.openiddict.com/errors/ID2095"
      }.
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext was successfully processed by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+ProcessChallengeErrorResponse`1[[OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext, OpenIddict.Validation, Version=3.1.1.0, Culture=neutral, PublicKeyToken=35a561290d20de2f]].
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext was marked as handled by OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandlers+ProcessChallengeErrorResponse`1[[OpenIddict.Validation.OpenIddictValidationEvents+ProcessChallengeContext, OpenIddict.Validation, Version=3.1.1.0, Culture=neutral, PublicKeyToken=35a561290d20de2f]].
info: OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandler[13]
      AuthenticationScheme: OpenIddict.Validation.AspNetCore was forbidden.
```

I find this strange because my I debugged my authorization policies and the conditions seem to be correct (at least the scope assertion does). I don't know where this error comes from but it seems to be something that OpenIddict throws (because of the error format).

I'm using the following configuration for:
Authorization server: https://gist.github.com/Thodor12/eab711b9e2d5cfbc96616a1e31911945
Resource server: https://gist.github.com/Thodor12/0a4bd83157d3470590e61e0989c56b67
Authorization endpoint: https://gist.github.com/Thodor12/d6d65143e9b19335668525c2b2142ece

I have populated my scopes through code, not in the database, my endpoints use the Authorize attribute like so: `[Authorize(Policy = nameof(Scope.Customers))]`.

I also have a small middleware to unpack the scopes (so that `mobile.access` is transformed into `mobile.access customers` etc (`mobile.access` is a combination of other scopes).

Let me know if you need any more info or see something out of the ordinary, I'm a bit stumped on the problem.

**Comment by kevinchalet** [comment]
Hey,

> Hello, first off thanks for making this amazing library.

Thanks for your kind words 😄 

> I don't know where this error comes from but it seems to be something that OpenIddict throws (because of the error format).

The line `AuthenticationScheme: OpenIddict.Validation.AspNetCore was forbidden` present in the logs - thanks for including them! - indicates that the 403 error response is likely not triggered by OpenIddict itself but by something else in your pipeline (that called `IAuthenticaticationService.ForbidAsync()`). In most cases, `ForbidAsync()` is indirectly called by the authorization middleware.

Sadly, the authorization middleware is not very good at logging useful stuff so I'd start by removing your custom authorization policy to see if the error disappears or not.

**Comment by Thodor12** [comment]
If I remove the `[Authorize]` attribute entirely (the one I mentioned) it still does not work.

My base controller does have `[Authorize(AuthenticationSchemes = OpenIddictValidationAspNetCoreDefaults.AuthenticationScheme)]` could that have anything to do with it?

**Comment by kevinchalet** [comment]
> My base controller does have `[Authorize(AuthenticationSchemes = OpenIddictValidationAspNetCoreDefaults.AuthenticationScheme)]` could that have anything to do with it?

It's unlikely: 403 responses indicate an authorization issue that can't be resolved by (re)authenticating. If there was an authentication issue, you'd get a 401 response.

Can you please add the following event handler in the OpenIddict validation configuration in your API project?

```csharp
options.AddEventHandler<OpenIddictValidationEvents.ProcessChallengeContext>(builder =>
{
    builder.SetOrder(int.MinValue);

    builder.UseInlineHandler(context =>
    {
        var trace = new StackTrace().ToString();

        return default;
    });
});
```

Then put a breakpoint on the `return default;` line and grab the stack trace from the `trace` variable. This should help figure out what's triggering it.

**Comment by Thodor12** [comment]
```
   at NextLevel.API.Extensions.AuthorizationExtensions.<>c.<AddAndConfigureAuthorization>b__0_6(ProcessChallengeContext context)
   at OpenIddict.Validation.OpenIddictValidationDispatcher.DispatchAsync[TContext](TContext context)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at OpenIddict.Validation.OpenIddictValidationDispatcher.DispatchAsync[TContext](TContext context)
   at OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandler.HandleChallengeAsync(AuthenticationProperties properties)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at OpenIddict.Validation.AspNetCore.OpenIddictValidationAspNetCoreHandler.HandleChallengeAsync(AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.ForbidAsync(AuthenticationProperties properties)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.ForbidAsync(AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationService.ForbidAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Authentication.AuthenticationService.ForbidAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ForbidAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ForbidAsync(HttpContext context, String scheme)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
```
There's more lines below this but those don't seem relevant.
Also, I'm still testing this currently with the authorization policy commented out.

**Comment by kevinchalet** [comment]
```csharp
   at Microsoft.AspNetCore.Authentication.AuthenticationService.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ChallengeAsync(HttpContext context, String scheme)
```

It's super weird, the authorization middleware triggers a challenge operation even though authentication seemed successful (in OpenIddict, `Challenge` and `Forbid` are assimilated to the same operation concept and OpenIddict automatically choses between 401 and 403 depending on whether the authentication was initially successful or not).

Try to remove `[Authorize(...)]` and add `var result = await HttpContext.AuthenticateAsync(OpenIddictValidationAspNetCoreDefaults.AuthenticationScheme)` in one of your actions and take a look at the properties of `AuthenticateResult` to see if there's a principal returned or an error.

**Comment by kevinchalet** [comment]
Ah, I see you edited your message, so my first remark no longer applies (did you forget to attach the token? 😄)

**Comment by Thodor12** [comment]
![2022-08-04 21_24_05-THOM-PC_SQLEXPRESS IPM_Dev - dbo CRM_Company - Microsoft SQL Server Management S](https://user-images.githubusercontent.com/14359461/182935988-ecb70187-9277-4651-abe7-0edf8edb128d.png)
My tokens do get attached to the requests yes

**Comment by kevinchalet** [comment]
> My tokens do get attached to the requests yes

I meant before you edited your message.

Your updated logs include the following lines:

```csharp
   at Microsoft.AspNetCore.Authentication.AuthenticationService.ForbidAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ForbidAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ForbidAsync(HttpContext context, String scheme)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
```

This indicates that for some reasons, one of the authorization policies enforced for the request failed and `ForbidAsync()` was called by the authorization middleware. It's definitely not OpenIddict that is triggering this errored response.

Do you see anything related to authorization in the logs?

**Comment by Thodor12** [comment]
I don't have any other logging but everything from OpenIddict, do you know of any way I can debug the authorization policy so I can check if all the requirements pass?

I did notice the following
![2022-08-04 21_33_06-IPMWebsite (Debugging) - Microsoft Visual Studio](https://user-images.githubusercontent.com/14359461/182937544-f96dca1e-bd93-4f71-92a2-f4a72acaf229.png)

The IsInRole check claims I'm not part of the given role even though the claims do include the role claim, shouldn't the identity have this role set?

**Comment by kevinchalet** [comment]
> The IsInRole check claims I'm not part of the given role even though the claims do include the role claim, shouldn't the identity have this role set?

If this check doesn't work, it's likely because the `ClaimsIdentity` doesn't have its `RoleClaimType` correctly configured.
In your claims transformer, do you re-create the `ClaimsIdentity`? If so, do you specify the correct claim type in the constructor?

**Comment by Thodor12** [comment]
The claims transformer returns the original principal passed to it, only modifying the scopes.
The identity says that the RoleClaimType is `http://schemas.microsoft.com/ws/2008/06/identity/claims/role`, does it have to be this exact string?

**Comment by kevinchalet** [comment]
It's not the correct value, it should be `role`. Can you share your transformer?

**Comment by Thodor12** [comment]
```cs
public class ScopeClaimResolver : IClaimsTransformation
{
    public Task<ClaimsPrincipal> TransformAsync(ClaimsPrincipal principal)
    {
        var scopes = new HashSet<string>();
        foreach (var scope in principal.GetScopes().ToHashSet())
        {
            scopes.Add(scope);
            ExtractScopes(scopes, scope);
        }

        principal.SetScopes(scopes);
        return Task.FromResult(principal);
    }

    private static void ExtractScopes(HashSet<string> scopes, string scope)
    {
        var scopesByKey = ApplicationScopes.GetScopesByKey();
        var scopesByName = ApplicationScopes.GetScopesByName();

        if (scopesByName.ContainsKey(scope))
        {
            var childScopes = scopesByName[scope].ChildScopes
                .Where(w => scopesByKey.ContainsKey(w))
                .Select(s => scopesByKey[s].ScopeName)
                .ToList();
            foreach (var childScope in childScopes)
            {
                scopes.Add(childScope);
                ExtractScopes(scopes, childScope);
            }
        }
    }
}
```

**Comment by Thodor12** [comment]
Also here's what the Identity looks like in total:
![2022-08-04 21_50_02-+ Actor = null](https://user-images.githubusercontent.com/14359461/182940304-14320575-d514-4aec-b1ee-77016cb0cda5.png)

(It's what I get from `assertion.User` in the policy)

**Comment by Thodor12** [comment]
Oddly enough, the AuthenticationType says `Server`, I'd expect `Validation` since this is the resource server, is that correct?

**Comment by kevinchalet** [comment]
Ahhhhhhh, that's because you use the opt-in ASP.NET Core Data Protection integration (there's an issue in 3.x where the OpenIddict validation handler will allow the DP integration to validate tokens locally even if you enabled introspection but it should be fixed in 4.x: I'll consider backporting this fix to 3.x).

Unlike JWTs, Data Protection tokens preserve the `AuthenticationType`, `NameClaimType` and `RoleClaimType` properties you set server-side. Since you use `new ClaimsIdentity(OpenIddictServerAspNetCoreDefaults.AuthenticationScheme)`, the identity will use the default name/role claims (the WS-Fed versions, not the OIDC ones) and these values will be preserved in the DP token. API-side, when validating the tokens, these properties are restored but they don't match the name of the "role" claims you use (i.e `role`).

TL;DR: replace `new ClaimsIdentity(OpenIddictServerAspNetCoreDefaults.AuthenticationScheme)` by `new ClaimsIdentity(TokenValidationParameters.DefaultAuthenticationType, Claims.Name, Claims.Role)` and it should work. I'll fix the 4.0 preview samples to use this overload so that things work flawlessly when using the opt-in ASP.NET Core Data Protection integration.

**Comment by Thodor12** [comment]
You're a lifesaver, I would not have figured that one out 😄, it works like a charm!

**Comment by kevinchalet** [comment]
BTW, you can remove `options.AddEncryptionCredentials(encryptionKey);` and `options.UseDataProtection();` as they should have no effect when using introspection.

**Comment by Thodor12** [comment]
I'm not 100% on what everything does, introspection asks the authorization server if the token they received is valid or what does it do?

**Comment by kevinchalet** [comment]
> introspection asks the authorization server if the token they received is valid

Yes, that's exactly what it does (by sending a backchannel HTTP request)

**Comment by Thodor12** [comment]
Alright, good to know, saves me the trouble of having to throw that certificate around to multiple runtimes, haha.
Thanks so much for your help, I'll close this now.

**Comment by kevinchalet** [comment]
Glad I could help 😃

**Comment by Thodor12** [comment]
Hey, I ran into the bug again but for a different reason now. I'm using the introspection endpoint now so my resource server cannot decode my access token anymore, however the claims identity does not get any scopes anymore. This is because the introspection endpoint doesn't give back any scopes so how would the resource server be able to validate if the token has any of the required scopes to call an endpoint?

This is the only thing that the introspection returns:
```json
{
  "active": true,
  "iss": "https://localhost:7068/",
  "sub": "5",
  "token_type": "Bearer",
  "token_usage": "access_token",
  "client_id": "mobile",
  "iat": 1660114571,
  "nbf": 1660114571,
  "exp": 1660118171
}
```

**Comment by kevinchalet** [comment]
Hey,

For security reasons, OpenIddict will return a stripped-down introspection response in two cases:
  - The caller is a public client.
  - The caller is not explicitly listed as an audience for the token.

You'll likely want to double-check the client_id of your introspection client is listed when calling `principal.SetResources()`.

Cheers.

**Comment by Thodor12** [comment]
I added audencies to my resource server and authorization server but now it's erroring saying that the authorized party is not valid:
```
warn: OpenIddict.Server.OpenIddictServerDispatcher[0]
      The introspection request was rejected because the access token was issued to a different client or for another resource server.
dbug: OpenIddict.Server.OpenIddictServerDispatcher[0]
      The event OpenIddict.Server.OpenIddictServerEvents+ValidateIntrospectionRequestContext was successfully processed by OpenIddict.Server.OpenIddictServerHandlers+Introspection+ValidateAuthorizedParty.
dbug: OpenIddict.Server.OpenIddictServerDispatcher[0]
```

Authorization endpoint:
```cs
        var identity = new ClaimsIdentity(OpenIddictServerAspNetCoreDefaults.AuthenticationScheme, Claims.Name, Claims.Role);
        var principal = new ClaimsPrincipal(identity)
            .SetClaim(Claims.Audience, "ipm-api")
            .SetClaim(Claims.Subject, userId)
            .SetClaim(Claims.Email, await _userManager.GetEmailAsync(user))
            .SetClaim(Claims.Name, await _userManager.GetUserNameAsync(user))
            .SetClaims(Claims.Role, (await _userManager.GetRolesAsync(user)).ToImmutableArray());
```

Resource server:
```cs
        services.AddOpenIddict()
            .AddValidation(options =>
            {
                options.SetIssuer(configuration.GetValue<string>("OpenId:Authority"));
                options.AddAudiences("ipm-api");

                options.UseIntrospection()
                    .SetClientId(configuration.GetValue<string>("OpenId:IntrospectionClientId"))
                    .SetClientSecret(configuration.GetValue<string>("OpenId:IntrospectionClientSecret"));

                options.UseSystemNetHttp();
                options.UseAspNetCore();
            });
```

**Comment by kevinchalet** [comment]
As I said in my last post, you must set the resources using `principal.SetResources()`:

> You'll likely want to double-check the client_id of your introspection client is listed when calling `principal.SetResources()`.

**Comment by Thodor12** [comment]
Sorry, I added "api" to `principal.SetResource()` but now I'm getting the error:
```
info: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The response was successfully returned as a challenge response: {
        "error": "missing_token",
        "error_description": "The security token is missing.",
        "error_uri": "https://documentation.openiddict.com/errors/ID2000"
      }.
```

However I can see just prior in the logging:
```
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+HandleIntrospectionResponseContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandlers+Introspection+PopulateClaims.
trce: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The token 'CfDJ8OpP9CaZdrhJkH68OTXlDjyxLfMygosTBD8KR6LPeIEsrDKq8XXvkJ9hqz0O8zwWDlnknjX8XeymEJ3Q65V6u5-GS2bBF3lpQLCxqwmrZWjUm7qMyULSm0s8R4lrksE2yU3yKPPCw9RJ_vKpYGPhOxYLEZNBZJUoIjTdKIyES8A-HU3BONJ8xZPOHxIgru-Ap0QGc88ysfKt17_jCDh-MbFaBG53mnruCRX06eJCXZG7Q_p3lUyI_5Cyi5Lkf_DM_Oro2_VRp--YLVeVHszKAiP3U4ptXcLYY-fu5ddcTaa7QUv_gKmuyfH_1A1BLQrrPA9vbxuE_xXsZOJXHImqayJqQylUlPK0b-0iJrbvILQWKqg5KxXYkVQ_DN89HksEpNl2tK7Bl7W9ATsVs2aAHQlOG24EZGlHfcf-RvWwLngMhtcWbkVrds7JUw-bbE1ngI1JKTQ3PW0me44uAwftrTuOIra6yNQAkiM61xHjqm-sIKh_C_CZjRdeSEZJ7CXfPZFso0g4hGhvt380S83WV0TWlhxiT7W9fsNmwTRz1ff_Q2Ue4YmP9xY2bmsAvNDwOL6cNz-8FsH3Jrr4d1w1jzi0hk1St6_4pGomxpuPD8Q3h8KyaAP6AJ54nmGCH-jDwFM2aq7wuTw2XePq4nqSJtDDxzKXj31Y2NaR8YyhcMlu_dWu77aNTqTKYApZ7LnhCMBz9FUUIomS1tEG2ryuKirTo5g_v0KrqfJNMfVI_z7Q' was successfully introspected and the following claims could be extracted: username: Admin, sub: 5, scope: openid profile email roles offline_access mobile.access, client_id: mobile, iat: 1660137667, exp: 1660141267, aud: api, email: thom.vandenakker@outlook.com, name: Admin, role: Admin, oi_tkn_typ: access_token.
```

**Comment by kevinchalet** [comment]
I tested locally with the latest OpenIddict 3.1.1 bits and I wasn't able to reproduce this issue.
Is your application open-source? If it's not, a repro app would help.

**Comment by Thodor12** [comment]
I can add you to the repo, maybe you can see something I don't

**Comment by Thodor12** [comment]
I sent an invite, there's 2 projects, NextLevel.Auth and NextLevel.API, those are the authorization server and resource server respectively.

Both projects have the OpenIddict configuration inside AuthorizationExtensions.cs

**Comment by kevinchalet** [comment]
Well, these projects are not exactly "minimal" and need some heavy configuration to run properly 😄 
I managed to execute the migrations but I'm eventually blocked by a `HttpContext does not contain any tenant information` exception during the login flow (likely because I end up with a blank database).

If you want me to take a closer look, please provide a minimal repro.

**Comment by Thodor12** [comment]
I'll see if I can put together a minimal repo later 👍🏻

**Comment by Thodor12** [comment]
I tried to replicate the problem by forking the Zirku example and modifying that, but to no avail.

I tried testing in my application a bit before but I'm noticing something strange in the logs.
![2022-08-10 20_00_46-](https://user-images.githubusercontent.com/14359461/183984215-4be9c7cd-550b-45b3-b8fd-559068c1b4d5.png)

Those 3 red critical errors are logged by myself through a custom event:
```cs
                options.AddEventHandler<ProcessAuthenticationContext>(builder =>
                {
                    builder.SetOrder(int.MaxValue);

                    builder.UseInlineHandler(context =>
                    {
                        context.Logger.LogCritical("NEW REQUEST BORDER");
                        context.Logger.LogCritical(context.Token);
                        context.Logger.LogCritical(context.TokenType);
                        return default;
                    });
                });
```

The odd thing is, I see 2x a ProcessAuthenticationContext, 1x for the OpenIddictValidationHandler and another for the OpenIddictValidationAspNetCoreHandlers, why does 1 of them get the token but the other one doesn't?

Also why does the AspNetCore even attempt to fetch the token again when the context already got one in the first place?

Am I not allowed to use `options.UseAspNetCore();` in the resource server?

**Comment by Thodor12** [comment]
It looks like these are separate requests sent in quick succession, so the logs are probably confused.
This is only happening on my mobile app, not the website, that one turns out to work fine. So I think I have to look at my client side application to see what it's doing.

**Comment by kevinchalet** [comment]
> Am I not allowed to use `options.UseAspNetCore();` in the resource server?

You have to call it. Otherwise, it won't work at all as the ASP.NET Core integration won't be registered if you don't.

You're likely on the right track: the logs likely include traces for 2 concurrent requests, with one of them not including the access token and being rejected for this reason.

**Comment by Thodor12** [comment]
> > Am I not allowed to use `options.UseAspNetCore();` in the resource server?
> 
> You have to call it. Otherwise, it won't work at all as the ASP.NET Core integration won't be registered if you don't.
> 
> You're likely on the right track: the logs likely include traces for 2 concurrent requests, with one of them not including the access token and being rejected for this reason.

Hey, after some further debugging it doesn't seem to be 2 independent requests, I've fixed some problems in my mobile app but the `missing_token` problem keeps occuring, one request seems to go all the way through the logs that I showed, it's 1 complete request.

Full log for a singular request here: https://gist.github.com/Thodor12/3895ddbb97d7bca40de7548bd11b63a3
This log is repeated twice in the logs for 2 requests.

I have at least seen that it's not an issue on my client side as I've put a logger onto my local requests now, and it shows:
```
 LOG  /customers Bearer CfDJ8OpP9CaZdrhJkH68OTXlDjz6AxrBz1VX4z3v--3wqFRLG99fW1yF-aaZ59Rh3r-fEyCTwUV7UBvdWs5X3_EPh2S34-DuIxht5JXprNTULFZXcislYKF3wUycmCW4-8278uBh8C-y0BuZ5YI_kuMbcqCZEOIBqhkXA_lXB0GigD9S5Hpl5iOkGlgYm0wEzB0qjPPxhUXJw0QpF2UzON6t-P-KXkb5Og4UN_9sw_EQU9OmUXqe0rV-1GgjggIbvAx88PfzeTq2Lb2dsQspySRoiNNBvGO0jvLqJvGuaCxEyBOV3mEjqnP8nK9te3WBlIvFvBoHcoewG_4B9WZ5284-_UzFSV4aDBwu-86uiFDOhP-9zvjdWDa_UYRs71EEOf3GCTWJtLf_E2DRLQS5x6a0_UGRer-uYRbb_Ml7359moYrACwaUmyQRw26nKoVw7Ls-Ce2vruUkYL7DuOiT9RN5nhrTTCjXIM6A1ZdArkUs7gPrQ-qBMULPt7EvdMjf6pZz1MaWuGvecgKdqyAA4zskArissxHhFZ5crYd0FuvQYBnNBdhWh3Fk3HtRNnG_7SYtnPcE0mfC7Q7a4x8ntWcrTVrxePUmao-2u3zW4J94paXm3m4OQnP9a_orZOUqMqKYRwCm9ACz-mKLf9QdUXNFRQCfngzgrtsepLoOvlS2yIqgr565KtQ9MCyxbP17HodoOSS2r7Oef8lyDTiUbUsKZSWS75y3yTMuMFZ2nG3C59fD
 LOG  [AxiosError: Request failed with status code 401]
 LOG  /customers Bearer CfDJ8OpP9CaZdrhJkH68OTXlDjwzGaW4FsuZtq2e8RGFbxthhYCJYvw88aNl5xFRGMS6oHjVZd5zVOWsTYuDcg9fczDQjECyLvXYSyZvm-ig428ofVwSNs8icSAeNZTq9HrFYTNrtpFvq4vC-ad50fp5O-DC-ec7cMp-J18XuBngDQoebbWv7aZTZLVNXYxn-pCjinLAN_qZE-cPpgAn8XdoUhvtxCOySjG8tuEFJXYIAnqqO7H5YBRSiPu_BJqG_6AUPW0rbLWszdDzk0izsHpo4UPMlrmhVE3BxVJY9P7_Z432HXHJERgROHwCK1u6xK7tdfG4-dYpjzUwQD_L-xG_aXVMurfC7cYbN5EfUVfpZobz128kVRxGwFpMA9zfut14ikfg9SqA_tQ7o8nhAEWLYjIY6BtxESkGXnw_-d-RE8CPKXP-DQrubIIEiWnrSZ-8OBVqpXQFsY3eHZCyDLxnoHenHYSKk0dWuLPASL3MSrcHq3G_4rNIm1Fk-_202vKoG5IlcMdwWZVysOOxojz1RQx4IjmNQFsVnK79L1ULB10vjRXbU4Yf7LHDJt3uVTKqwTrwlLXMvKYRS-KOPz3bRZPn5u1DVrDA_B92ZbqI6sP771Z_qd8T8RpuuWSTvUshRPajyWavQCQ5E-ErzQZeOtZADGE8yMSf6o9xMUCY3_22QSM2CR5JJoVNelzcoX3Per_mL7iuO0lvpVVjMm1aD1mtggiVWvT-Zmak3DikQS8c
 ```
 
The request seems to be getting rejected for a wrong reason.
 
I checked the source and it seems the `ValidateRequiredTokens` step is throwing the error, that one throws the `missing_token` error. According to the logs just prior to that runs a step `ExtractAccessTokenFromAuthorizationHeader`.
I've also checked the code for that one and it does look like the token is present as the authorization header too:
```cs
                options.AddEventHandler<ProcessAuthenticationContext>(builder =>
                {
                    builder.SetOrder(int.MaxValue);

                    builder.UseInlineHandler(context =>
                    {
                        context.Logger.LogCritical(context.Transaction.GetHttpRequest().Headers.Authorization);
                        return default;
                    });
                });
```
```
crit: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
Bearer CfDJ8OpP9CaZdrhJkH68OTXlDjwzGaW4FsuZtq2e8RGFbxthhYCJYvw88aNl5xFRGMS6oHjVZd5zVOWsTYuDcg9fczDQjECyLvXYSyZvm-ig428ofVwSNs8icSAeNZTq9HrFYTNrtpFvq4vC-ad50fp5O-DC-ec7cMp-J18XuBngDQoebbWv7aZTZLVNXYxn-pCjinLAN_qZE-cPpgAn8XdoUhvtxCOy
```

Do you have any idea why the transaction could just seemingly randomly lose the AccessToken?

(Also how can I get my custom event handler to be closer to the handler which rejects it, so I can put a debugger in between there?)

(I'm very sorry that I have to ask so much of you but I'm at a bit of a loss)

**Comment by kevinchalet** [comment]
> Do you have any idea why the transaction could just seemingly randomly lose the AccessToken?

It's actually not exactly the OpenIddict transaction but the attached ASP.NET Core `HttpContext` that seems to have a null `Authorization` header in the second part of the logs. And in this case, since the header is not present, OpenIddict does the right thing when it rejects the request.

To be frank, the logs you shared don't make much sense to me at all: the `ProcessAuthentication(Context)` event should never been run twice for the same request (it's _in theory_ impossible as the ASP.NET Core authentication stack internally caches the task returned by [`HandleAuthenticateAsync()`](https://github.com/openiddict/openiddict-core/blob/3.1.1/src/OpenIddict.Validation.AspNetCore/OpenIddictValidationAspNetCoreHandler.cs#L110), that itself caches the context instance in the OI transaction anyway).

Can you please include the rest of the ASP.NET Core logs, including the HTTP request logs?

> I checked the source and it seems the `ValidateRequiredTokens` step is throwing the error, that one throws the missing_token error.

It's worth noting this handler is new in OpenIddict 4.0, where the authentication pipeline has been muchly improved to be able to let handlers determine what type of tokens are to be validated (mostly to deal with non-standard stuff 😄). The equivalent in 3.1 is `ValidateToken`: https://github.com/openiddict/openiddict-core/blob/02c64fdd90d5d5d54e918b508088f3dce74233fb/src/OpenIddict.Validation/OpenIddictValidationHandlers.cs#L54-L96

> (Also how can I get my custom event handler to be closer to the handler which rejects it, so I can put a debugger in between there?)

The `Order` (set via `SetOrder`) controls when the handler is invoked: if you want your custom handler to be invoked prior to a built-in one, make sure you use a lower order (it works pretty much like MVC filters if you're familiar with them)

**Comment by Thodor12** [comment]
> Can you please include the rest of the ASP.NET Core logs, including the HTTP request logs?

My god I think I finally see the problem, I'm using localtunnel to be able to actually access the API on the application I'm running on my phone, and I saw this in the logs:
```
crit: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      Bearer CfDJ8OpP9CaZdrhJkH68OTXlDjxlg_whxnzQNe0IBibhyR2JBYWLuLAvAmZqDrHhT_n0_z890H-Rmi_zV33I--YggAlTbWT7Lo7vHYAaG2MOQUItBjop6dGYnMoaGTj5ETQv1IOXogeyvYIrfWrATjtNTBPN1jI-e4DIxMAXYzfuxzCIrMPxczIf1Ijp3Ac2GXwcdwAZwWctM5QlGKU8SEcYL_rumym_GlhpMPAco78RcxsHxqq7XIxJwk8bu9m0ZDsgB9iahIp9jI-IwHHOc3sHe85joeErPniqF3xtP0ERujLdO2-1v5_Sue5Bil6USKEudoRfO-kYE5ENM1Xp9fSrjk9nkhaUARDn8WIE5XlQsHpOGCVyuqXXb_R3qvahEEtu_jDcXlujDWvIh9Jd-Z0kP329AnvZPTgVuTRRGk1IwykH7I3UdDkRVTkgyvtOlBwA8LpaK-q_N4gVfdsrHWXKwrN0K7NdSJw1cYbIwb6ak3xd2uytBqnS8K4aCFPplEgjiIGqqlhrrn3DW503Vk9IDHXIvzymYakFiEVjW6Mn3rNhYrR40_v36jiiZKjYTJRqxESJugQAqYh4VfYU69Y1Q6OSDvqpANOurcNnsXiyu3YuYpoEhlrzcahTvRdm2pSR5sE5HggKqpuYmJXiBZncCEDGOkvbkI1dwebzXNjBpP0Fqjet7s1cSXejXB6W4KXgLBtZkzFkOlEQAtoDlcO6doHCBS0mdqy5n1zybhKrYwq5
dbug: OpenIddict.Validation.OpenIddictValidationDispatcher[0]
      The event OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext was successfully processed by OpenIddict.Validation.OpenIddictValidationHandler`1[[OpenIddict.Validation.OpenIddictValidationEvents+ProcessAuthenticationContext, OpenIddict.Validation, Version=3.1.1.0, Culture=neutral, PublicKeyToken=35a561290d20de2f]].
info: Microsoft.AspNetCore.HttpLogging.HttpLoggingMiddleware[2]
      Response:
      StatusCode: 302
      Location: https://ipm-api.loca.lt/api/customers
info: Microsoft.AspNetCore.HttpLogging.HttpLoggingMiddleware[1]
      Request:
      Protocol: HTTP/1.1
      Method: GET
      Scheme: http
      PathBase:
      Path: /api/customers
      Accept: application/json, text/plain, */*
      Connection: close
      Host: ipm-api.loca.lt
      User-Agent: okhttp/4.9.2
      Accept-Encoding: gzip
      x-forwarded-port: [Redacted]
      x-forwarded-ssl: [Redacted]
      x-forwarded-proto: [Redacted]
      x-forwarded-for: [Redacted]
      x-real-ip: [Redacted]
      x-forwarded-host: [Redacted]
```
It's doing a https redirect to the localtunnel domain, that's probably why it loses it's token...

I feel sorry for blaming the problem on OpenIddict, I didn't think this could've been the problem, I'm just gonna go rework that redirect configuration and bash localtunnel once more to stop the redirect funkiness.

At least you still managed to help me find the problem so thank you so much for that!
I'll upgrade the sponsor a level since this has basically become a second support request within the first one 😄

**Comment by kevinchalet** [comment]
> It's doing a https redirect to the localtunnel domain, that's probably why it loses it's token...

Ockham would love that 🤣 

> I feel sorry for blaming the problem on OpenIddict, I didn't think this could've been the problem

No problem 😄 

> At least you still managed to help me find the problem so thank you so much for that!
I'll upgrade the sponsor a level since this has basically become a second support request within the first one 😄

Glad I could help! Thanks for sponsoring the project 👍🏻
